opm/simulators/linalg/bda/BdaBridge.cpp

```diff
@@ -0,0 +1,242 @@
+template <class BridgeMatrix>
+void getSparsityPattern(BridgeMatrix& mat, std::vector<int> &h_rows, std::vector<int> &h_cols) {
+	int sum_nnzs = 0;
+
+	// convert colIndices and rowPointers
+	if(h_rows.size() == 0){
+		h_rows.emplace_back(0);
+			for(typename BridgeMatrix::const_iterator r = mat.begin(); r != mat.end(); ++r){
+				int size_row = 0;
+				for(auto c = r->begin(); c != r->end(); ++c){
+					h_cols.emplace_back(c.index());
+					size_row++;
+				}
+				sum_nnzs += size_row;
+				h_rows.emplace_back(sum_nnzs);
+			}
+		// set last rowpointer
+		h_rows[mat.N()] = mat.nonzeroes();
+	}
+} // end getSparsityPattern()
+
+#endif
+
+template <class BridgeMatrix, class BridgeVector>
+void BdaBridge::solve_system(BridgeMatrix *mat OPM_UNUSED, BridgeVector &b OPM_UNUSED, InverseOperatorResult &res OPM_UNUSED)
+{
+
+#if HAVE_CUDA
+	if(use_gpu){
+		BdaResult result;
+		result.converged = false;
+		static std::vector<int> h_rows;
+		static std::vector<int> h_cols;
+		int dim = (*mat)[0][0].N();
+		int N = mat->N()*dim;
+		int nnz = mat->nonzeroes()*dim*dim;
+
+		if(dim != 3){
+			OpmLog::error("cusparseSolver only accepts blocksize = 3 at this time");
+			use_gpu = false;
+		}
+
+		if(h_rows.capacity() == 0){
+			h_rows.reserve(N+1);
+			h_cols.reserve(nnz);			
+#if PRINT_TIMERS_BRIDGE
+			Dune::Timer t;
+#endif
+			getSparsityPattern(*mat, h_rows, h_cols);
+#if PRINT_TIMERS_BRIDGE
+			std::ostringstream out;
+			out << "getSparsityPattern() took: " << t.stop() << " s";
+			OpmLog::info(out.str());
+#endif
+		}
+
+#if PRINT_TIMERS_BRIDGE
+		Dune::Timer t_zeros;
+		int numZeros = checkZeroDiagonal(*mat);
+		std::ostringstream out;
+		out << "Checking zeros took: " << t_zeros.stop() << " s, found " << numZeros << " zeros";
+		OpmLog::info(out.str());
+#else
+		checkZeroDiagonal(*mat);
+#endif
+
+
+		/////////////////////////
+		// actually solve
+
+		typedef cusparseSolverBackend::cusparseSolverStatus cusparseSolverStatus;
+		// assume that underlying data (nonzeroes) from mat (Dune::BCRSMatrix) are contiguous, if this is not the case, cusparseSolver is expected to produce garbage
+		cusparseSolverStatus status = backend->solve_system(N, nnz, dim, static_cast<double*>(&(((*mat)[0][0][0][0]))), h_rows.data(), h_cols.data(), static_cast<double*>(&(b[0][0])), result);
+		switch(status){
+		case cusparseSolverStatus::CUSPARSE_SOLVER_SUCCESS:
+			//OpmLog::info("cusparseSolver converged");
+			break;
+		case cusparseSolverStatus::CUSPARSE_SOLVER_ANALYSIS_FAILED:
+			OpmLog::warning("cusparseSolver could not analyse level information of matrix, perhaps there is still a 0.0 on the diagonal of a block on the diagonal");
+			break;
+		case cusparseSolverStatus::CUSPARSE_SOLVER_CREATE_PRECONDITIONER_FAILED:
+			OpmLog::warning("cusparseSolver could not create preconditioner, perhaps there is still a 0.0 on the diagonal of a block on the diagonal");
+			break;
+		default:
+			OpmLog::warning("cusparseSolver returned unknown status code");
+		}
```

Comment:
BdaBridge now receives a cusparseSolverStatus and prints a warning if something is wrong.